### ISSUE TITLE
Support different array backends

### ISSUE THREAD

**Issue by Oisin-M** [ecmwf/earthkit-hydro#117]
### Is your feature request related to a problem? Please describe.

It could be interesting to support different array backends, since we mostly just use functions like
`np.add.at(out, indices, values)`

This is not part of the array backend standard, but there are equivalents to this in most backends:
```
cupy.add.at(out, indices, values) # cupy, in-place, potentially not deterministic
out.scatter_add_(dim, indices, values) # torch, in-place, potentially not deterministic
jax.numpy.add.at(out, indices, values) # jax, returns new array, deterministic
tf.tensor_scatter_nd_add(tensor, indices, updates) # tensorflow, returns new array, potentially not deterministic
```
Annoyingly, some are not in-place. Some are also not deterministic but that's not much of an issue.

The reason this could be very interesting is for potential GPU use, but also because all operations in ekh would be ***differentiable***, so we could have machine learning models use directly the functions and backprop after e.g. you could do a convolutional layer over a full catchment, and then have a loss function based on the upstream average.

Potential problem: The fast rust code for the topological labels computation is entirely numpy-dependent at the moment, but we could always convert

**Comment by Oisin-M** [comment]
Added in v1.0.0
